xmtp_mls/src/storage/sql_key_store.rs

```diff
@@ -22,60 +32,923 @@ impl<'a> SqlKeyStore<'a> {
     pub fn conn(&self) -> &DbConnection<'a> {
         self.conn
     }
+
+    pub fn write<const VERSION: u16>(
+        &mut self,
+        label: &[u8],
+        key: &[u8],
+        value: &[u8],
+    ) -> Result<(), <Self as StorageProvider<CURRENT_VERSION>>::Error> {
+        let storage_key = build_key_from_vec::<VERSION>(label, key.to_vec());
+        let query = "REPLACE INTO storage (storage_key, version, data) VALUES (?, ?, ?)";
+
+        self.conn.raw_query(|conn| {
+            sql_query(query)
+                .bind::<diesel::sql_types::Binary, _>(&storage_key)
+                .bind::<diesel::sql_types::Text, _>(&VERSION.to_string())
+                .bind::<diesel::sql_types::Binary, _>(&value)
+                .execute(conn)
+        });
+
+        Ok(())
+    }
+
+    pub fn append<const VERSION: u16>(
+        &mut self,
+        label: &[u8],
+        key: &[u8],
+        value: &[u8],
+    ) -> Result<(), <Self as StorageProvider<CURRENT_VERSION>>::Error> {
+        let storage_key = build_key_from_vec::<VERSION>(label, key.to_vec());
+        let query = "INSERT INTO storage (storage_key, version, data) VALUES (?, ?, JSON_ARRAY_APPEND(COALESCE((SELECT data FROM storage WHERE storage_key = ? AND version = ?), '[]'), '$', ?)) ON DUPLICATE KEY UPDATE data = VALUES(data)";
```

Comment:
Does SQLite actually support `ON DUPLICATE KEY`?

I thought [it didn't](https://stackoverflow.com/questions/23622504/on-duplicate-key-not-working-in-sqlite). But they do have an alternative

I'm also not sure if [SQLite supports `JSON_ARRAY_APPEND`](https://www.sqlite.org/json1.html)

Reply:
I don't think reading and then writing is the end of the world if there is no alternative to `json_array_append`